### PR TITLE
OpenAI-DotNet 7.3.0

### PR DESCRIPTION
**Pull request by StephenHodgson**
- Added `AgentsEndpoint`
- Added `ThreadsEndpoint`
- Updated `ImagesEndpoint` return types to `ImageResult` list
- Updated `FilesEndpoint.ListFilesAsync` with optional `purpose` filter query parameter.
- Refactored list responses with a more generic `ListQuery` and `ListResponse<TObject>` pattern
  - `EventList` -> `ListResponse<EventResponse>`
  - `FineTuneJobList` -> `ListResponse<FineTuneJobResponse>`
- Standardized names for timestamps to have suffix: `UnixTimeSeconds`
- Standardized response class names (existing classes depreciated)
  - `FileData` -> `FileResponse`
  - `CompletionResult` -> `CompletonResponse`
  - `Event` -> `EventResponse`
  - `FineTuneJob` -> `FineTuneJobResponse`

### CODE REVIEW

**Comment by LucGosso** [comment]
Looking good so far!

File: OpenAI-DotNet-Tests/TestFixture_03_Chat.cs [hunk 1]
```diff
@@ -418,6 +418,7 @@ public async Task Test_07_GetChatToolCompletion()
             };
 
             var chatRequest = new ChatRequest(messages, tools: tools, toolChoice: "auto");
+            OpenAIClient.ChatEndpoint.EnableDebug = true;
```
Comment (StephenHodgson) [inline OpenAI-DotNet-Tests/TestFixture_03_Chat.cs:421]:
```suggestion
```

File: OpenAI-DotNet/Assistants/AssistantsEndpoint.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
+using OpenAI.Extensions;
+using OpenAI.Files;
+using System;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI.Assistants
+{
+    public class AssistantsEndpoint : BaseEndPoint
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Assistants/AssistantsEndpoint.cs:10]:
```suggestion
    public sealed class AssistantsEndpoint : BaseEndPoint
```

File: OpenAI-DotNet/Threads/ThreadsEndpoint.cs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI.Threads
+{
+    /// <summary>
+    /// Create threads that assistants can interact with.<br/>
+    /// <see href="https://platform.openai.com/docs/api-reference/threads"/>
+    /// </summary>
+    public class ThreadsEndpoint : BaseEndPoint
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Threads/ThreadsEndpoint.cs:13]:
```suggestion
    public sealed class ThreadsEndpoint : BaseEndPoint
```

File: OpenAI-DotNet/FineTuning/FineTuneJob.cs [hunk 4]
```diff
@@ -4,7 +4,8 @@
 
 namespace OpenAI.FineTuning
 {
-    public sealed class FineTuneJob
+    [Obsolete("use FileTuneJobResponse")]
```
Comment (StephenHodgson) [inline OpenAI-DotNet/FineTuning/FineTuneJob.cs:7]:
```suggestion
    [Obsolete("use FineTuneJobResponse")]
```

File: OpenAI-DotNet/FineTuning/FineTuneJobList.cs [hunk 5]
```diff
@@ -1,8 +1,10 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using System.Text.Json.Serialization;
 
 namespace OpenAI.FineTuning
 {
+    [Obsolete("Use ListResponse<FineTuneJob>")]
```
Comment (StephenHodgson) [inline OpenAI-DotNet/FineTuning/FineTuneJobList.cs:7]:
```suggestion
    [Obsolete("Use ListResponse<FineTuneJobResponse>")]
```

File: .github/workflows/Publish-Nuget.yml [hunk 6]
```diff
@@ -31,24 +31,39 @@ env:
 jobs:
   build:
     if: ${{ !github.event_name == 'pull_request' || !github.event.pull_request.draft }}
-    runs-on: windows-latest
+    runs-on: linux-latest
```
Comment (StephenHodgson) [inline .github/workflows/Publish-Nuget.yml:34]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/Publish-Nuget.yml [hunk 7]
```diff
@@ -31,24 +31,39 @@ env:
...
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
       with:
         fetch-depth: 0
+
     - uses: actions/setup-dotnet@v3
       with:
         dotnet-version: ${{ env.DOTNET_VERSION }}
+
     - uses: microsoft/setup-msbuild@v1
```
Comment (StephenHodgson) [inline .github/workflows/Publish-Nuget.yml:45]:
```suggestion
```

File: .github/workflows/Publish-Nuget.yml [hunk 8]
```diff
@@ -31,24 +31,38 @@ env:
...
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
       with:
         fetch-depth: 0
+
     - uses: actions/setup-dotnet@v3
       with:
         dotnet-version: ${{ env.DOTNET_VERSION }}
-    - uses: microsoft/setup-msbuild@v1
+
```
Comment (StephenHodgson) [inline .github/workflows/Publish-Nuget.yml:44]:
```suggestion
```

File: .github/workflows/Publish-Nuget.yml [hunk 9]
```diff
@@ -31,24 +31,38 @@ env:
...
     - name: Test Packages
-      run: dotnet test --configuration Release
       if: ${{ github.ref != 'refs/heads/main' && github.event_name != 'push' }}
+      run: dotnet test --configuration Release --verbosity normal --collect:"XPlat Code Coverage" --results-directory ./coverage
       env:
         OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
         OPENAI_ORGANIZATION_ID: ${{ secrets.OPENAI_ORGANIZATION_ID }}
 
+    - name: Code Coverage Summary Report
+      if: ${{ github.ref != 'refs/heads/main' && github.event_name != 'push' }}
```
Comment (StephenHodgson) [inline .github/workflows/Publish-Nuget.yml:54]:
```suggestion
      if: ${{ github.ref != 'refs/heads/main' && github.event_name != 'push' && always() }}
```

File: OpenAI-DotNet/Threads/CreateThreadRunRequest.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
+using OpenAI.Assistants;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json.Serialization;
+
+namespace OpenAI.Threads
+{
+    public sealed class CreateThreadRunRequest
+    {
+        public CreateThreadRunRequest(AssistantResponse assistant, string model = null, string instructions = null, IEnumerable<Tool> tools = null, IReadOnlyDictionary<string, string> metadata = null)
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply OpenAI-DotNet/Threads/CreateThreadRunRequest.cs]: seems this comment was from an old change a few days ago. This is no longer the case.

File: OpenAI-DotNet/Threads/CreateThreadRunRequest.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            AssistantId = assistant.Id;
+            Model = string.IsNullOrWhiteSpace(model) ? Models.Model.GPT3_5_Turbo : model;
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply OpenAI-DotNet/Threads/CreateThreadRunRequest.cs]: Great catch! I'll update this accordingly.

File: OpenAI-DotNet/Threads/CreateThreadAndRunRequest.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        /// The ID of the assistant to use to execute this run.
+        /// </summary>
+        [JsonPropertyName("assistant_id")]
+        public string AssistantId { get; }
+
+        /// <summary>
+        /// Thread
+        /// </summary>
+        [JsonPropertyName("thread")]
+        public ThreadResponse ThreadResponse { get; }
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply OpenAI-DotNet/Threads/CreateThreadAndRunRequest.cs]: I believe the latest version accurately reflects this.

If you believe it is still a problem, submit a new review on the latest changes please.

**Pull request merged**